src/librbd/object_map/RefreshRequest.cc

```diff
@@ -133,6 +133,10 @@ Context *RefreshRequest<I>::handle_load(int *ret_val) {
     return nullptr;
   } else if (*ret_val < 0) {
     lderr(cct) << "failed to load object map: " << oid << dendl;
+    if (*ret_val == -ETIMEDOUT) {
```

Comment:
You will need to invalidate the object-map at this point (if it's for the HEAD revision)

Reply:
backgroud info: in iSCSI scenario, set `rados_osd_op_timeout = 10`
1. create a test image `lun1`
2. client A opens and writes to image `lun1`
3. client B opens image `lun1`
4. use `tc` to add a delay to a storage node C
5. client A write timed out and close image `lun1` 
6. client B try to acquire exclusive lock and failed (the object-map of `lun1` happens to be at node C)
7. client B successfully acquired exclusive lock, but the object-map is set to invalid

we think in this scenario, object-map should not set to invalid and return error to client B.
not sure if this is the right way to reslove it.

@dillaman  what do you think? thanks.
```
2020-03-19 11:13:15.292862 7f60177fe700  5 librbd::ManagedLock: 0x7f600800ac30 handle_acquire_lock: successfully acquired exclusive lock
2020-03-19 11:13:15.292868 7f60177fe700 10 librbd::ExclusiveLock: 0x7f600800ac30 post_acquire_lock_handler: r=0
2020-03-19 11:13:15.292876 7f60177fe700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 send_open_object_map:
2020-03-19 11:13:15.292885 7f60177fe700 20 librbd::object_map::RefreshRequest: 0x7f6010004570 send: object_count=524288
2020-03-19 11:13:15.292891 7f60177fe700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 send_lock: oid=rbd_object_map.1afb6b8b4567
2020-03-19 11:13:15.292894 7f60177fe700 10 librbd::object_map::LockRequest: 0x7f6010001ac0 send_lock: oid=rbd_object_map.1afb6b8b4567
2020-03-19 11:13:16.294808 7f6017fff700 10 librbd::object_map::LockRequest: 0x7f6010001ac0 handle_lock: r=0
2020-03-19 11:13:16.294821 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_lock
2020-03-19 11:13:16.294827 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 send_load: oid=rbd_object_map.1afb6b8b4567
2020-03-19 11:13:26.294961 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_load: r=-110
2020-03-19 11:13:26.294979 7f6017fff700 -1 librbd::object_map::RefreshRequest: failed to load object map: rbd_object_map.1afb6b8b4567
2020-03-19 11:13:26.294987 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 send_invalidate
2020-03-19 11:13:26.327480 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 invalidating object map in-memory
2020-03-19 11:13:26.327487 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 invalidating object map on-disk
2020-03-19 11:13:36.327620 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 should_complete: r=-110
2020-03-19 11:13:36.327635 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_invalidate: r=0
2020-03-19 11:13:36.327647 7f6017fff700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 handle_open_object_map: r=0
```

Reply:
If you cannot open the object map, that means you won't update the object-map, which means that it needs to be invalidated. If your OSDs are timing out after 10 seconds, perhaps object-map is the least of your concerns.

Why, in your example, does the load time-out when client B opens the object-map?

Reply:
@dillaman  If time-out occurs when open the objectmap and return error code to APPs, we have reason to suspect that there may be a problem with the network, but if time-out and returns to normal, we donot necessarily know that it is caused by the network.

I think we should try to avoid setting the object map to invalid. There are two main considerations: first, many RBD operations are based on the object map, which will be time-consuming and inefficient without it; second, it may take a long time to rebuild a large image (eg. 10TB) which will affect the business.

In addition, if invalidating failed, the opened image will run in the no object map state, but at the same time it can continue writing. Actually, object map on the server is still valid, which will cause the mismatch between the object map and the actual image object and may cause image damage.

> Why, in your example, does the load time-out when client B opens the object-map?

Cause the object map of `lun1` happens to be at node C which its network is in sub-health state (we have this test case).

Thank you.

Reply:
We shouldn't return an error to the app on IO -- and I don't see that in your appended log section:

> 2020-03-19 11:13:36.327647 7f6017fff700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 handle_open_object_map: r=0

Reply:
> In addition, if invalidating failed, the opened image will run in the no object map state, but at the same time it can continue writing. Actually, object map on the server is still valid, which will cause the mismatch between the object map and the actual image object and may cause image damage.

Now, that's a potential issue but the only solution would be to return IO errors in that case (or keep retrying to invalidate until successful if you hit a `ETIMEDOUT` error)